packages/common/src/store/pages/saved-page/reducer.ts

```diff
@@ -21,19 +29,106 @@ import { signOut } from 'store/sign-out/slice'
 import { ActionsMap } from 'utils/reducer'
 
 import { PREFIX as tracksPrefix } from './lineups/tracks/actions'
-import { LibraryCategory, SavedPageState } from './types'
+import { LibraryCategory, LibraryCategoryType, SavedPageState } from './types'
 
 const initialState = {
   // id => uid
-  localSaves: {},
-  saves: [],
+  localTrackFavorites: {},
+  localTrackPurchases: {},
+  localTrackReposts: {},
+  localAlbumFavorites: [],
+  localAlbumPurchases: [],
+  localAlbumReposts: [],
+  localRemovedAlbumFavorites: [],
+  localRemovedAlbumReposts: [],
+  localPlaylistFavorites: [],
+  localPlaylistReposts: [],
+  localPlaylistPurchases: [],
+  localRemovedPlaylistFavorites: [],
+  localRemovedPlaylistResposts: [],
+  trackSaves: [],
   initialFetch: false,
   hasReachedEnd: false,
   fetchingMore: false,
   tracks: initialLineupState,
   selectedCategory: LibraryCategory.Favorite
 } as SavedPageState
 
+/** Utility to get the name of key in which locally added or removed collections are stored in SavedPageState.
+ * For example, playlists that are favorited/unfavorited locally go into `localPlaylistFavorites`/`localPlaylistRemovedFavorites`.
+ */
+const getLocalCollectionStateKeys = ({
+  isAlbum,
+  type
+}: {
+  isAlbum: boolean
+  type: Omit<LibraryCategoryType, 'All'>
+}) => {
+  let categoryKeySuffix
+  if (type === 'favorite') {
+    categoryKeySuffix = 'Favorites'
+  } else if (type === 'repost') {
+    categoryKeySuffix = 'Reposts'
+  } else if (type === 'purchase') {
+    categoryKeySuffix = 'Purchases'
+  }
+  const additionKey = (
+    isAlbum
+      ? `localAlbum${categoryKeySuffix}`
+      : `localPlaylist${categoryKeySuffix}`
+  ) as keyof SavedPageState
+  const removalKey = (
+    isAlbum
+      ? `localRemovedAlbum${categoryKeySuffix}`
+      : `localRemovedPlaylist${categoryKeySuffix}`
+  ) as keyof SavedPageState
+  return { additionKey, removalKey }
+}
+
+const removeCollectionLocally = ({
+  state,
+  action,
+  type
+}: {
+  state: SavedPageState
+  action: { collectionId: ID; isAlbum: boolean }
+  type: Omit<LibraryCategoryType, 'All'>
+}) => {
+  const { additionKey, removalKey } = getLocalCollectionStateKeys({
+    isAlbum: action.isAlbum,
+    type
+  })
+  return {
+    ...state,
+    [additionKey]: (state[additionKey] as ID[]).filter(
+      (id) => id !== action.collectionId
+    ),
+    [removalKey]: [action.collectionId, ...(state[removalKey] as ID[])]
+  }
+}
+
+const addCollectionLocally = ({
+  state,
+  action,
+  type
+}: {
+  state: SavedPageState
+  action: { collectionId: ID; isAlbum: boolean }
+  type: 'repost' | 'favorite'
+}) => {
+  const { additionKey, removalKey } = getLocalCollectionStateKeys({
+    isAlbum: action.isAlbum,
+    type
+  })
+  return {
+    ...state,
```

Comment:
eek, really wish this was using `createSlice`. It feels error-prone to make sure we're correctly spreading state in all of these actions.

Reply:
100%